### ISSUE TITLE
Conditional GDScript parsing

### ISSUE THREAD

**Issue by wkubiak** [godotengine/godot#6340]
**Operating system or device - Godot version:**
Godot 2.1 stable

**Issue description** (what happened, and what was expected):
Like we have preprocessor statements in C or other languages where we can notify the compiler that such and such code should be compiled if a condition is true or not (like am I compiling on Windows or Linux for example), it would be nice to have such conditional parsing in GDScript.

Since 2.1 we have in GDScript the nice function "type_exists(string)", which returns a boolean value if a given type exists or not in our Godot instance.

The problem is, I can't use this tool for anything other than checking just that.

If I'd put that method call in an IF statement, the code under true/false conditions is still parsed by Godot and obviously this will result in an error because I write something regarding a non-existent type.

Thankfully Godot can just ignore the script file in which I write such statements, but still I have to make special arrangements just for that (like keeping all my platform-specific or module-specific code in separate scripts).

This is of course just a single example use-case.

Or maybe there is a way to force conditional parsing of which I'm not aware of? I asked around on the IRC channels but got no answers in this regards, so I suppose that it's not possible to do such a thing currently.

**Comment by neikeq** [comment]
I don't like the C preprocessor. It's a completely different language on top of C. If we are going to add compile time code generation I would prefer to do it the D way: https://dlang.org/pretod.html

BTW, scripts are not compiled in debug mode (IINW). This means the preprocessor would need to run every time you run your project which could be slow.

**Comment by wkubiak** [comment]
@neikeq I just used the C preprocessor as a comparison of functionality. It doesn't have to be done the exact same way. The "D way" of handling such situations looks nicer of course.

**Comment by SleepProgger** [comment]
While i'd love some Conditional GDScript parsing, i guess the mentioned problem could be solved if the ObjectTypeDB::instance function would be exposed to GDscript.  
This way one could do:
```
if type_exists("SomeClass"):
  var some_instance = instance("SomeClass")
```
This would make it easy to use a module class if the module is available and fall back to a GDscript version if not.

**Comment by vnen** [comment]
@SleepProgger you can already do that using the standard way of instancing (by calling `.new()`):

```GDScript
if type_exists("SomeClass"):
    var some_instance = SomeClass.new()
```

**Comment by SleepProgger** [comment]
@vnen You'd get a parser error if SomeClass does not exists.

**Comment by vnen** [comment]
@SleepProgger you are correct. I've seen [this pattern applied before](https://github.com/Geequlim/godot-code/blob/a8771429088999bb0a50d48f559c82f049527122/addons/tiled_importer/map.gd#L73-L82) so I assumed it would work.

**Comment by bojidar-bg** [comment]
What about having (again D-way a bit) a parser-level `const if`?

```gdscript
func stuff():
  const if type_exists("SomeClass"): # if not, won't really parse contents, probably
    add_child(SomeClass.new())
  else: # No need of const here, but would be needed for elif I think
    show_unsupported_dialog("SomeClass")
```

**Comment by neikeq** [comment]
`static if` to be more precise. Although @SleepProgger's problem could perhaps be solved by improving the parser to detect the `type_exists` condition.

Here is a dirty trick you can use for now to avoid this specific problem:

``` GDScript
# someclass_factory.gd

static func get_new():
    return SomeClass.new()

# whatever.gd

var SomeClassFactory = null setget , get_someclass_factory

func get_someclass_factory():
    # must lazy initialize in order to work
    if SomeClassFactory == null:
        SomeClassFactory = load("someclass_factory.gd")
    return SomeClassFactory

func _ready():
    if type_exists("SomeClass"):
        var someclass_instance = SomeClassFactory.get_new()
```

From 3.0 there is an even better and not hackish solution if all you want is to create a new instance (what else would you need it for? :stuck_out_tongue:):

``` GDScript
func _ready():
    if ClassDB.can_instance("SomeClass"):
        var someclass_instance = ClassDB.instance("SomeClass")
```

**Comment by neikeq** [comment]
To clarify, I am not sure if `static if` would not be a good idea. It would complicate static analysis (like code completion). I think a simple language like GDScript would do well just with the `version` keyword.

**Comment by RandomShaper** [comment]
Without knowing this was being discussed, I opened another issue. I've closed it and I'm pasting here what I proposed:

# Optimizing out `OS.is_debug_build()`

I tend to put assertions here and there in my GDScripts. Sometimes just not a plain assert but a call to a more complex routine to check if certain preconditions are met:

```
func _ready():
    if OS.is_debug_build():
        debug_check_nodes_z():
```

Do you think that letting the compiler handle `OS.is_debug_build()`  as a constant could provide some benefit in terms of performance or something?

**Comment by mhilbrunner** [comment]
As there does not seem to be much interest in this and above improvements are now merged, as well as with 3.0 making some of this easier in other ways, vote for closing, @neikeq?

**Comment by wkubiak** [comment]
You are right. Issue seems to be irrelevant already because of the merged improvements.
Closing issue.
